<launch>

<arg name="playback"     default="true"/>
<arg name="visualizar"   default="true" />
<arg name="sync_odom"    default="true"/>

<arg name="camera_frame"  default="base_link"/>
<arg name="optical_frame" default="camera_color_left"/>
<arg name="base_frame"    default="base_link" />
<arg name="odom_frame"    default="world"     />

<!--Fazendo nuvem de pontos a partir de imagens na ocasiao de tocando bags-->
<group if="$(arg playback)">
<!--    Tocando o bag tambem-->
    <include file="$(find kitti_datasets)/launch/playback.launch">
    </include>

    <group ns="stereo">
    <node name="stereo_image_proc" pkg="stereo_image_proc" type="stereo_image_proc" output="screen">
        <remap from="/stereo/left/image_raw"    to="/kitti/left/image_raw"/>
        <remap from="/stereo/right/image_raw"   to="/kitti/right/image_raw"/>
        <remap from="/stereo/left/camera_info"  to="/kitti/left/camera_info"/>
        <remap from="/stereo/right/camera_info" to="/kitti/right/camera_info"/>

        <param name="approximate_sync" value="true"/>
        <param name="queue_size" value="10"/>
        <!--Melhorar visao stereo -->
        <param name="prefilter_size" value="77"/>
        <param name="prefilter_cap" value="34"/>
        <param name="disparity_range" value="50"/>
        <param name="min_disparity" value="0"/>
        <param name="correlation_window_size" value="35"/>
        <param name="speckle_size" value="1000"/>
        <param name="speckle_range" value="3"/>
        <param name="fullDP" value="true"/>
        <param name="uniqueness_ratio" value="1.0"/>
        <param name="texture_threshold" value="1000"/>
    </node>
    </group>
</group>

<!--lancando o VISO2 com as configuracoes necessarias-->
<node pkg="viso2_ros" type="stereo_odometer" name="stereo_odometer" output="screen">
    <remap from="stereo" to="/kitti"/>
    <remap from="image" to="image_raw"/>
    <remap from="camera_info" to="camera_info"/>
    <param name="sensor_frame_id" value="/naosei"/>
    <param name="base_link_frame_id" value="$(arg base_frame)"/>
    <param name="approximate_sync" value="true"/>
    <param name="publish_tf" value="true"/>
    <param name="queue_size" value="10"/>
    <param name="odom_frame_id" value="$(arg odom_frame)"/>

    <!--Ajustar melhores valores para fazer o match das nuvens-->
    <param name="max_features" value="2"/>
    <param name="bucket_width" value="50"/>
    <param name="bucket_height" value="=50"/>
    <param name="nms_n" value="3"/>
    <param name="nms_tau" value="50"/>
    <param name="match_binsize" value="50"/>
    <param name="match_radius" value="300"/> <!--esse parametro e bastante importante, afeta a distancia entre features-->
    <param name="outlier_disp_tolerance" value="5"/>
    <param name="outlier_flow_tolerance" value="5"/>
    <param name="match_disp_tolerance" value="5"/>
    <param name="half_resolution" value="1"/> <!--melhor esse mesmo-->
    <param name="refinement" value="1"/>
    <param name="ransac_iters" value="300"/>
    <param name="inlier_threshold" value="1.5"/>
    <param name="reweighting" value="true"/> <!--default ja-->
    <param name="ref_frame_change_method" value="1"/>
    <param name="ref_frame_motion_threshold" value="5.0"/>
    <param name="ref_frame_inlier_threshold" value="150"/>

</node>

<!--visualizar os resultados no rviz-->
<node if="$(arg visualizar)" pkg="rviz" type="rviz" name="rviz_orbslam" args="-d $(find duo3d_slam_pkg)/config/viso2_vis.rviz">
</node>

<!--caso de testar com o ground truth para avaliar odometria-->
<group if="$(arg sync_odom)">
    <node name="sincronizar_ground_truth" pkg="kitti_datasets" type="sync_dados" output="screen">
        <param name="image_topic"  value="/kitti/left/image_raw"/>
        <param name="gt_topic"     value="/kitti/oxts/gps/fix"/>
        <param name="camera_topic" value="/stereo_odometer/odometry"/>
        <param name="raio_terra"   value="3931000"/>
    </node>
</group>

</launch>
